subsys/bluetooth/host/scan.c

```diff
@@ -765,154 +872,285 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
-			 * more to come. This means the Controller is aborting the reassembly. We
-			 * discard the partially received report, and the application is not
-			 * notified.
-			 *
-			 * See the Controller's documentation for possible reasons for aborting.
-			 * Hint: CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX.
-			 */
-			LOG_DBG("Discarding incomplete advertisement.");
-			reset_reassembling_advertiser();
-			goto cont;
-		}
+	bool reassembling = r->state != FRAG_ADV_INACTIVE;
+	bool same_chain = advertisers_equal(r, &evt->addr, evt->sid);
 
-		if (is_new_advertiser) {
-			/* We are not reassembling reports from an advertiser and
-			 * this is the first report from the new advertiser.
-			 * Initialize the new advertiser.
-			 */
-			__ASSERT_NO_MSG(reassembling_advertiser.state == FRAG_ADV_INACTIVE);
-			init_reassembling_advertiser(&evt->addr, evt->sid);
-		}
+	LOG_DBG("[%s] status %s", bt_addr_le_str(&evt->addr), status_to_str(data_status));
 
-		if (evt->length + ext_scan_buf.len > ext_scan_buf.size) {
-			/* The report does not fit in the reassemby buffer
-			 * Discard this and future reports from the advertiser.
-			 */
-			reassembling_advertiser.state = FRAG_ADV_DISCARDING;
-		}
+	if (complete && !same_chain) {
+		/* Interleaved "complete" reports are allowed (for some reason). */
+		return false;
+	}
 
-		if (reassembling_advertiser.state == FRAG_ADV_DISCARDING) {
-			if (!more_to_come) {
-				/* We do no longer need to keep track of this advertiser as
-				 * all the expected data is received.
-				 */
-				reset_reassembling_advertiser();
-			}
-			goto cont;
-		}
+	if (reassembling && !same_chain) {
+		/* Drop interleaved "more-data" reports. We return "true" to
+		 * indicate that we have processed the report, but we silently
+		 * drop it.
+		 */
+		LOG_DBG("dropping interleaved fragmented report");
+		return true;
+	}
 
-		net_buf_simple_add_mem(&ext_scan_buf, buf->data, evt->length);
-		if (more_to_come) {
-			/* The controller will send additional reports to be reassembled */
-			continue;
-		}
+	if (!reassembling && !invalid) {
+		/* Try to allocate a backing buffer in which to reassemble data
+		 * from the fragmented advertising reports. If the chain is
+		 * invalid, we don't bother with allocation as we will discard
+		 * the chain anyways.
+		 */
+		init_reassembly(r, &evt->addr, evt->sid);
+	}
+
+	if (r->report == NULL) {
+		LOG_DBG("No reassembly buf.");
+		start_discarding(r);
+	} else if (evt->length > net_buf_tailroom(r->report)) {
+		LOG_DBG("Adv report doesn't fit reassembly buffer.");
+		start_discarding(r);
+	}
 
-		/* No more data coming from the controller.
-		 * Create event.
+	bool discarding = r->state == FRAG_ADV_DISCARDING;
+
+	if (!discarding) {
+		__ASSERT_NO_MSG(r->report);
+		net_buf_add_mem(r->report, /* flex the member 🤖💪 */ evt->data, evt->length);
+	}
+
+	if (!discarding && complete) {
+		/* Core Spec v5.4. 4.E.7.7.65.13 says controller uses latest
+		 * RSSI. Host wants to be like controller when grow up.
```

Comment:
> Host wants to be like controller when grow up.

Can you elaborate on what that means?

Reply:
Hey, I'll comment with my personal account :)

@alwa-nordic should rephrase it. Basically, it means that the host should not have a surprising behavior: if the spec has a guideline for the controller, the host should also (unless otherwise stated) try to follow it.

Here the guideline for the LL is that it should use the RSSI from the latest AUX packet received for enqueuing the HCI ext-adv-report event. In the host we translate that to "use the RSSI value from the last ext-adv-report fragment in the chain when you notify the application".
I hope that makes sense.

Reply:
Ah, that makes sense.